Simplenote/src/main/java/com/automattic/simplenote/repositories/SimperiumTagsRepository.kt

```diff
@@ -48,4 +55,71 @@ class SimperiumTagsRepository(
             false
         }
     }
+
+    override suspend fun deleteTag(tag: Tag) {
+        deleteTagFromNotes(tag)
+        tag.delete()
+    }
+
+    private fun deleteTagFromNotes(tag: Tag) {
+        val cursor = tag.findNotes(notesBucket, tag.name)
+
+        while (cursor.moveToNext()) {
+            val note = cursor.getObject()
+            val tags = note.tags
+            tags.remove(tag.name)
+            note.tags = tags
+            note.save()
+        }
+
+        cursor.close()
+    }
+
+    @ExperimentalCoroutinesApi
+    override suspend fun tagsChanged(): Flow<Boolean> = callbackFlow {
+        val callbackOnSaveObject = Bucket.OnSaveObjectListener<Tag> { _, _ -> offer(true) }
+        val callbackOnDeleteObject = Bucket.OnDeleteObjectListener<Tag> { _, _ -> offer(true) }
+        val callbackOnNetworkChange = Bucket.OnNetworkChangeListener<Tag> { _, _, _ -> offer(true) }
+
+        tagsBucket.addOnSaveObjectListener(callbackOnSaveObject)
+        tagsBucket.addOnDeleteObjectListener(callbackOnDeleteObject)
+        tagsBucket.addOnNetworkChangeListener(callbackOnNetworkChange)
+        AppLog.add(AppLog.Type.SYNC, "Added tag bucket listener (TagsActivity)")
+
+        awaitClose {
+            tagsBucket.removeOnSaveObjectListener(callbackOnSaveObject)
+            tagsBucket.removeOnDeleteObjectListener(callbackOnDeleteObject)
+            tagsBucket.removeOnNetworkChangeListener(callbackOnNetworkChange)
+            AppLog.add(AppLog.Type.SYNC, "Removed tag bucket listener (TagsActivity)")
+        }
+    }
+
+    override suspend fun allTags(): List<TagItem> {
+        val tagQuery = Tag.all(tagsBucket).reorder().orderByKey().include(Tag.NOTE_COUNT_INDEX_NAME)
+        val cursor = tagQuery.execute()
+
+        return cursorToTagItems(cursor)
+    }
+
+    override suspend fun searchTags(query: String): List<TagItem> {
+        val tags = Tag.all(tagsBucket)
+                .where(Tag.NAME_PROPERTY, Query.ComparisonType.LIKE, "%$query%")
```

Comment:
note that here too we will want to compare not the tag name but the tag hash. the `name` for the tag in the Tag bucket might be a different representation than the name given in the query even though they are the same tag. if we don't do this then we'll miss tags in our searches that should appear.

Reply:
@dmsnell do you have some examples with I can work and test. I think I understand the problem but I would like some samples you have had problems with before so I can add tests to be sure we are covering those cases.

Reply:
We'll see if I can do this without GitHub mangling the text. (Edit: nope)

 - `Bu\u{0308}cher` and `B\u{00fc}cher` "Bücher" are the same tag
 - `q\u{0307}\u{0323}` and `q\u{0323}\u{0307}` "q̣̇" are the same tag

If these two cases are handled properly (even if it works with the query itself) then it's probably all fine. It _can_ be difficult to get the right test inputs since text entry systems do like to perform their own normalization (that is, even if copying and pasting, sometimes when you paste two different things they turn into the same thing, and it's hard to tell since they are both visually identical).

Reply:
Hi @dmsnell ,

I have tried both cases:

> Bu\u{0308}cher and B\u{00fc}cher "Bücher" are the same tag
![imagen](https://user-images.githubusercontent.com/81183210/123501107-33148f80-d608-11eb-8888-4612b840f99f.png)

 &nbsp;
 &nbsp; 
  

> q\u{0307}\u{0323} and q\u{0323}\u{0307} "q̣̇" are the same tag
![imagen](https://user-images.githubusercontent.com/81183210/123501041-ea5cd680-d607-11eb-9d2f-ce81af91cc7a.png)

It seems that both cases worked. However,

>  It can be difficult to get the right test inputs since text entry systems do like to perform their own normalization (that is, even if copying and pasting, sometimes when you paste two different things they turn into the same thing

Yes, this made me hesitate a little bit about the results, but let me add an issue to add tests for the `SimperiumTagsRepository` and allow me to investigate more how to format the inputs so we can test this properly (see issue #1403).

Reply:
Good checking! If it's too hard _and_ especially if this is already broken, let's not worry too much about it. Otherwise, in a test (or for a quick check by modifying the code) we should be able to manually add a new tag with a hard-coded name, as we can do that in Java strings as well.

```java
TagUtils.createTagIfMissing(tagBucket, "q\u0307\u0323");
TagUtils.createTagIfMissing(tagBucket, "q\u0323\u0307");
```

and then we can perform a tag search and inspect the tag bucket/database to see if it created two entities.